delivery_carrier_by_line/models/stock_move.py

```diff
@@ -0,0 +1,30 @@
+# Copyright 2021 Camptocamp (https://www.camptocamp.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import logging
+
+from odoo import models
+
+_logger = logging.getLogger(__name__)
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    def _domain_search_picking_for_assignation_default(self):
+        domain = super()._domain_search_picking_for_assignation_default()
+        domain.append(("carrier_id", "=", self.sale_line_id.carrier_id.id))
+        return domain
+
+    def _get_new_picking_values(self):
+        vals = super()._get_new_picking_values()
+        carrier = self.mapped("sale_line_id.carrier_id")
+        if len(carrier) > 1:
+            _logger.warning("Moves belongs to different carriers %s" % carrier)
```

Comment:
According to me, the call to super will already fail on
`vals['carrier_id'] = self.mapped('sale_line_id.order_id.carrier_id').id`

I would move that test before the call to super and raise an error instead of logging a warning.
The else case is managed by the call to super

Reply:
the parent method check for carrier on order, here I am checking for carrier on order lines, these two checks are not equal